crypto/rand/rand_unix.c

```diff
@@ -128,7 +128,7 @@ int RAND_poll(void)
 #   error "librandom not (yet) supported"
 #  endif
 
-int RAND_poll(void)
+int RAND_poll_ex(RAND_poll_cb cb, void *arg)
```

Comment:
Just wondering if the `ok++` statements could be replaced by jump to cleanse and return success.  This won't make any difference if only one source is defined but it will if several are.  It seems reasonable to define both `OPENSSL_RAND_SEED_RDCPU` and `OPENSSL_RAND_SEED_DEVRANDOM` when building for x86 platforms.  The rdcpu path checks for the capability on the CPU and would fail if the HRNG wasn't supported.

If one path is executed successfully, it means enough random bytes have been collected to satisfy the current request, so an early exit makes sense.  Using multiple sources, however, is safer in the sense that one being compromised (e.g. hardware failure) doesn't cause an overall failure.  Multiple sources will require more resources however -- in the example above, rdcpu will be very fast whereas opening and reading files will be much slower (& will block sometimes).

The resource versus security trade off could be delegated to the user via a configuration option although I fear the code would get messy by doing so.  Given that the user is specifying the randomness sources, the user is implicitly trusting them and the short circuit would be a reasonable option.


If this is done `OPENSSL_RAND_SEED_GETRANDOM` should be moved before `OPENSSL_RAND_SEED_DEVRANDOM` because `getrandom(2)` is preferable to reading device files and both are asking the kernel for the bytes.  Maybe do this anyway so the more desirable of the two is first.